Add GitHub integration to bug cards (#3671)

satvikpathak opened the pull request:
## Summary
This PR implements GitHub integration for bug cards, displaying GitHub issue links, status, and comment counts.

## Changes Made
- Added `github_state` and `github_comment_count` fields to the `Issue` model  
- Added `fetch_github_data()` method to retrieve GitHub issue data via API  
- Updated bug card template to display GitHub information  
- Added database migration for new fields  
- Styled using Tailwind CSS with full dark mode support (no inline styles)

## Features
- **GitHub Link**: Clickable link with icon (opens in new tab)  
- **Status Badge**: Color-coded (green = open, red = closed)  
- **Comment Count**: Displays the number of comments on the GitHub issue  
- **Conditional Rendering**: Shown only if `github_url` is present  

## Testing
All code quality checks passed:
- Black formatting  
- isort import sorting  
- ruff linting  

## Migration Instructions
After merge, run:

```bash
python manage.py makemigrations
python manage.py migrate
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Issue pages now include a "View on GitHub" link with icon, a GitHub state badge (open/closed or unavailable), and a comment count.
  * UI shows clear fetch-status badges for GitHub data: loaded, not yet loaded, or data unavailable.

* **Chores**
  * Public IP timestamp field renamed from created to created_at — update any integrations or exports that depend on it.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

satvikpathak committed:
Add GitHub integration to bug cards (#3671)

- Add github_state and github_comment_count fields to Issue model
- Display GitHub link, status badge, and comment count on bug cards
- Add fetch_github_data() method to sync GitHub issue data via API
- Use Tailwind CSS for styling with dark mode support
- Include migration for new database fields

Closes #3671

satvikpathak committed:
Apply pre-commit formatting fixes

- Format template file with djLint
- Fix code formatting with ruff-format

satvikpathak committed:
Fix GitHub API authentication and URL parsing issues

- Change from 'token' to 'Bearer' format for modern GitHub tokens
- Use urlparse() for robust URL parsing (handles query strings/fragments)
- Clear stale github_state/github_comment_count on invalid URLs
- Validate path format (requires /owner/repo/issues/number)

Addresses CodeRabbit review feedback

satvikpathak committed:
Fix field consistency: Remove redundant null=True from github_comment_count

- Removed null=True from github_comment_count field (has default=0)
- Updated migration to match field definition
- Ensures consistent field behavior across model and migration

sidd190 commented on ISSUE_3671_IMPLEMENTATION.md:
```diff
@@ -0,0 +1,71 @@
+# Issue #3671: Link GitHub Issue on Bug Card
```
This can be removed. We probably don't want these files in the main branch.

sidd190 commented on website/models.py:
```diff
@@ -572,6 +572,8 @@ class Issue(models.Model):
     closed_by = models.ForeignKey(User, null=True, blank=True, related_name="closed_by", on_delete=models.CASCADE)
     closed_date = models.DateTimeField(default=None, null=True, blank=True)
     github_url = models.URLField(default="", null=True, blank=True)
+    github_state = models.CharField(max_length=10, null=True, blank=True)
```
Ahhh, consider using maybe a field `choices=[('open','Open'),('closed','Closed')]` here instead since we don't really want any state other than that.

sidd190 commented on website/templates/includes/_bug.html:
```diff
@@ -122,6 +122,42 @@
+                    </svg>
+                    <span>View on GitHub</span>
+                </a>
+                {% if bug.github_state %}
```
When applying migration, there may be many existing Issue records with no github_url. Their github_state will be null. We want the UI to handle that gracefully. This comes under the above comment, just consider to add handlers for both this and the above if conditions.

satvikpathak commented on website/models.py:
```diff
@@ -640,6 +642,61 @@ def get_cve_score(self):
+            parsed_url = urlparse(self.github_url)
+
+            # Validate it's a GitHub URL
+            if "github.com" not in parsed_url.netloc:
```
I’ll switch to validating with endswith() / allowed domains to avoid cases like github.com.evil.com. Thanks for pointing it out!

satvikpathak committed:
feat: Add GitHub state choices and fetch failure handling

- Add GITHUB_STATE_CHOICES with 'open' and 'closed' options for better maintainability
- Add github_data_fetch_failed boolean field to track API fetch failures
- Update github_state field to use choices instead of accepting any string
- Improve domain validation using endswith('.github.com') for security
- Add UX feedback with three display states:
  * Green badge with comment count when data successfully fetched
  * Yellow 'Data unavailable' badge when fetch fails
  * Gray 'Data not yet loaded' badge for legacy records
- Update fetch_github_data() to set failure flag and normalize states
- Create migration 0257 for new field and field choice updates

Addresses feedback from issue #3671 review

satvikpathak commented:
I did code changes for all comments. Going to see migrations one too

satvikpathak committed:
Merge migrations 0256 and 0257 into single 0256 migration for issue #3671

satvikpathak commented on website/models.py:
```diff
@@ -640,6 +642,61 @@ def get_cve_score(self):
             logger.warning(f"Error fetching CVE score for {self.cve_id}: {e}")
             return None
 
+    def fetch_github_data(self):
```
Thanks for pointing this out, this probably fits better as a management command. I’ll refactor it out of the model and add tests once I move it.

DonnieBLT replied:
We already have something like this

DonnieBLT commented on website/migrations/0256_add_github_fields_to_issue.py:
```diff
@@ -0,0 +1,36 @@
+        ),
+        migrations.AddField(
+            model_name="issue",
+            name="github_data_fetch_failed",
```
please make this more generic like github_fetch_result_data or github_fetch_status

DonnieBLT commented on website/templates/includes/_bug.html:
```diff
@@ -122,6 +122,42 @@
             {{ bug.url|truncatechars:60 }}
         </a>
+        <!-- GitHub Information -->
+        {% if bug.github_url %}
```
you can do this on your local setup.  Have you tested it locally?

satvikpathak committed:
Rename github_data_fetch_failed to github_fetch_status for issue #3671

github-actions[bot] committed:
Merge branch 'main' into fix/issue-3671-github-bug-card

DonnieBLT committed:
Merge branch 'main' into fix/issue-3671-github-bug-card

DonnieBLT committed:
Merge branch 'main' into fix/issue-3671-github-bug-card

satvikpathak committed:
refactor: use existing GitHub API utility and fix IP model index

- Refactored Issue.fetch_github_data() to use existing website.utils.fetch_github_data() utility
- Eliminated code duplication in GitHub API requests
- Fixed IP model index to reference 'created_at' field instead of 'created'
- Improved error handling for GitHub data fetching
- All code quality checks passed (Black, isort, ruff, pre-commit)

satvikpathak committed:
Merge branch 'main' into fix/issue-3671-github-bug-card

- Resolve merge conflict in models.py
- Add requests.exceptions.RequestException to exception handling
